### PR TITLE
[EC-357] - Move application gateway from core to modules

### PR DESCRIPTION
**Pull request by christian-calabrese**
### List of changes
Move `application gateway`'s related resources from `legacy core` folder to the new `common` one organizing files in modules, to be able to reuse them across two regions

### Motivation and context
As discussed several times, core TF configuration is too big and must be split through different configs. This file structure makes it easier reusing same definitions. Resources that will be included in this config are listed in Confluence RFC. Everything else will be moved apart

### Type of changes

- [ ] Add new resources
- [X] Update configuration to existing resources
- [ ] Remove existing resources

### Env to apply

- [ ] DEV
- [ ] UAT
- [X] PROD

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [X] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [ ] No

### Other information

---

### How to apply

After PR is approved
1. run deploy pipeline from Azure DevOps [io-platform-iac-projects](https://dev.azure.com/pagopaspa/io-platform-iac-projects/_build?view=folders)
2. select PR branch
3. wait for approval

### CODE REVIEW

**Review by mamu0 — APPROVED** [review]
LGTM

File: src/common/_modules/application_gateway/networking.tf [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Subnet to host the application gateway
+module "appgateway_snet" {
```
Comment (Krusty93) [inline src/common/_modules/application_gateway/networking.tf:2]: Can we remove the module and replace it with the resource? Mapping is 1:1

File: src/common/_modules/application_gateway/outputs.tf [hunk 2]
```diff
@@ -0,0 +1,3 @@
+output "public_ip" {
+  value = azurerm_public_ip.appgateway_public_ip.ip_address
```
Comment (Krusty93) [inline src/common/_modules/application_gateway/outputs.tf:2]: I suggest to add `id`, `name`, `resource_group_name` as well

File: src/common/_modules/application_gateway/variables.tf [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  description = "Azure region short name"
+}
+
+variable "tags" {
+  type        = map(any)
+  description = "Resource tags"
+}
+
+variable "resource_groups" {
+  type        = map(string)
```
Comment (Krusty93) [inline src/common/_modules/application_gateway/variables.tf:33]: I'm not sure about this variable structure. How can I know that a specific rg is there?

File: src/common/_modules/global/modules/dns/outputs.tf [hunk 4]
```diff
@@ -36,3 +36,7 @@ output "private_dns_zones" {
     servicebus = azurerm_private_dns_zone.privatelink_servicebus
   }
 }
+
+output "external_domain" {
+  value = var.external_domain
```
Comment (Krusty93) [inline src/common/_modules/global/modules/dns/outputs.tf:41]: no way to get it from resource's output?

File: src/common/_modules/application_gateway/variables.tf (hunk 3 above)
Comment (christian-calabrese) [reply src/common/_modules/application_gateway/variables.tf:33]: Good point! The thinking behind the choice is based on the lack (yet) of the resource groups structure we want to adopt during the migration to italy north

I kept it extremely free for this reason, but we can surely find a structure that can accomodate both regions and structures. Maybe changes in the usage of resource groups may be inconvenient, but I think it deserves a discussion

File: src/common/_modules/global/modules/dns/outputs.tf (hunk 4 above)
Comment (christian-calabrese) [reply src/common/_modules/global/modules/dns/outputs.tf:41]: It is used in several resources, I can't find a resource where we can get it straight away

This was put in place when the dns module was used in the core workspace.
After it has been moved to common, I'm torn whether we should remove the output and use a single variable in the common workspace. 

Using the output surely creates a stronger bond between the actual external domain and the resources that make use of it
What do you think?

File: src/common/_modules/application_gateway/outputs.tf [hunk 5]
```diff
@@ -0,0 +1,8 @@
+output "public_ip" {
+  value = {
+    address        = azurerm_public_ip.appgateway_public_ip.ip_address
+    id             = azurerm_public_ip.appgateway_public_ip.id
+    name           = azurerm_public_ip.appgateway_public_ip.name
+    resource_group = azurerm_public_ip.appgateway_public_ip.resource_group_name
```
Comment (Krusty93) [inline src/common/_modules/application_gateway/outputs.tf:6]:
```suggestion
    resource_group_name = azurerm_public_ip.appgateway_public_ip.resource_group_name
```

**Review by Krusty93 — APPROVED** [review]

**Pull request merged**